src/core/thread/mle.cpp

```diff
@@ -915,22 +915,23 @@ void Mle::SetRloc16(uint16_t aRloc16)
     if (aRloc16 != oldRloc16)
     {
         otLogNoteMle("RLOC16 %04x -> %04x", oldRloc16, aRloc16);
-
-        // Clear cached CoAP with old RLOC source
-        if (oldRloc16 != Mac::kShortAddrInvalid)
-        {
-            Get<Tmf::Agent>().ClearRequests(mMeshLocal16.GetAddress());
-        }
     }
 
-    Get<ThreadNetif>().RemoveUnicastAddress(mMeshLocal16);
+    if (Get<ThreadNetif>().HasUnicastAddress(mMeshLocal16) &&
```

Comment:
Just curious - do we need to execute any of this code if `aRloc16 == oldRloc16`? In other words, can we just have `VerifyOrExit(aRloc16 != oldRloc16)` at the top?

Reply:
Yes. I think the way we use these methods, we may need to run through this code even when `aRloc16 == oldRloc16`. 

The main reason is that `Mle` itself is not directly storing the RLOC16 value and `GetRloc16()` actually gets it from `Mac` using `Get<Mac::Mac>().GetShortAddress()`. It is possible that the MAC short address is set directly using `Mac::SetMacAddress()` without calling `SetRloc16()` (e.g., I think we do this in `Mle::Restore()` and then later call `SetRloc16(GetRloc16()))` from `Mle::Start()` to set up `mMeshLocal16` address, etc).